File: src/Ccmc.Core/Services/ShellMenuComposer.cs
namespace Ccmc.Core.Services;

/// <summary>One entry in the tray menu / jump list. Label is the folder name.</summary>
public sealed record ShellMenuEntry(string Label, string Path, bool IsPinned);

/// <summary>
/// Single source of truth for the tray menu and the taskbar jump list:
/// pinned projects first (config order), then recents (newest first) minus
/// anything already pinned, capped at <c>recentCap</c>.
/// </summary>
public static class ShellMenuComposer
{
    public static IReadOnlyList<ShellMenuEntry> Compose(
        IEnumerable<string> pinnedPaths, IEnumerable<string> recentPaths, int recentCap)
    {
        var entries = new List<ShellMenuEntry>();
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        foreach (var p in pinnedPaths)
        {
            if (string.IsNullOrWhiteSpace(p) || !seen.Add(p)) continue;
            entries.Add(new ShellMenuEntry(LabelOf(p), p, IsPinned: true));
        }

        var recents = 0;
        foreach (var p in recentPaths)
        {
            if (recents >= recentCap) break;
            if (string.IsNullOrWhiteSpace(p) || !seen.Add(p)) continue;
            entries.Add(new ShellMenuEntry(LabelOf(p), p, IsPinned: false));
            recents++;
        }
        return entries;
    }

    private static string LabelOf(string path) =>
        System.IO.Path.GetFileName(System.IO.Path.TrimEndingDirectorySeparator(path));
}

File: electron/src/core/launch/shellMenuComposer.ts
/**
 * Single source of truth for the tray menu and taskbar jump list:
 * pinned projects first (config order), then recents (newest first) minus
 * anything already pinned, capped at recentCap.
 *
 * Direct port of C# ShellMenuComposer. Pure — uses only `path` (allowed).
 */

import path from 'path'

/** One entry in the tray menu / jump list. Label is the folder name. */
export interface ShellMenuEntry {
  label: string
  path: string
  isPinned: boolean
}

export interface ComposeShellMenuOptions {
  pinnedPaths: readonly string[]
  recentPaths: readonly string[]
  recentCap: number
}

function labelOf(p: string): string {
  return path.basename(p.replace(/[\\/]+$/, ''))
}

/**
 * Composes a menu list from pinned + recent paths.
 * - Pinned paths appear first, in config order.
 * - Recent paths follow, newest-first, de-duped against pinned (case-insensitive).
 * - Total recents capped at recentCap.
 */
export function composeShellMenu(opts: ComposeShellMenuOptions): ShellMenuEntry[] {
  const { pinnedPaths, recentPaths, recentCap } = opts
  const entries: ShellMenuEntry[] = []
  const seen = new Set<string>()

  for (const p of pinnedPaths) {
    if (!p || p.trim().length === 0) continue
    const key = p.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    entries.push({ label: labelOf(p), path: p, isPinned: true })
  }

  let recents = 0
  for (const p of recentPaths) {
    if (recents >= recentCap) break
    if (!p || p.trim().length === 0) continue
    const key = p.toLowerCase()
    if (seen.has(key)) continue
    seen.add(key)
    entries.push({ label: labelOf(p), path: p, isPinned: false })
    recents++
  }

  return entries
}
